src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.cpp

```diff
@@ -25,42 +24,66 @@
 /// \cond
 
 namespace {
-std::array<DataVector, 9> compute_characteristic_speeds(
+void compute_characteristic_speeds(
+    const gsl::not_null<std::array<DataVector, 9>*> pchar_speeds,
     const Scalar<DataVector>& lapse, const tnsr::I<DataVector, 3>& shift,
     const tnsr::I<DataVector, 3>& spatial_velocity,
     const Scalar<DataVector>& spatial_velocity_squared,
     const Scalar<DataVector>& sound_speed_squared,
     const Scalar<DataVector>& alfven_speed_squared,
     const tnsr::i<DataVector, 3>& normal) noexcept {
-  auto characteristic_speeds =
-      make_array<9, DataVector>(-1.0 * get(dot_product(normal, shift)));
-
-  const auto auxiliary_speeds =
-      RelativisticEuler::Valencia::characteristic_speeds(
-          lapse, shift, spatial_velocity, spatial_velocity_squared,
-          Scalar<DataVector>{get(sound_speed_squared) +
-                             get(alfven_speed_squared) *
-                                 (1.0 - get(sound_speed_squared))},
-          normal);
-
-  characteristic_speeds[0] -= get(lapse);
-  characteristic_speeds[1] = auxiliary_speeds[0];
-  for (size_t i = 2; i < 7; ++i) {
-    // auxiliary_speeds[1], auxiliary_speeds[2], and auxiliary_speeds[3]
-    // are the same.
-    gsl::at(characteristic_speeds, i) = auxiliary_speeds[2];
+  auto& char_speeds = *pchar_speeds;
+  if (char_speeds[0].size() != get<0>(normal).size()) {
+    char_speeds[0] = DataVector(get<0>(normal).size());
+  }
+  Scalar<DataVector> temp0(char_speeds[0].data(), char_speeds[0].size());
+  dot_product(make_not_null(&temp0), normal, shift);
+  char_speeds[0] *= -1.0;
+  char_speeds[8] = char_speeds[0];
+
+  char_speeds[0] -= get(lapse);
+  // Mapping of indices between GRMHD char speeds and relativistic Euler char
+  // speeds arrays.
+  //
+  // GRMHD     Rel Euler
+  //   1           0
+  //   2           1
+  //   3           2
+  //   4           3
+  //   5           1
+  //   6           1
+  //   7           4
+  std::array<DataVector, 5> rel_euler_char_speeds{};
+  const size_t num_grid_points = get(lapse).size();
+  for (size_t i = 0; i < 5; ++i) {
+    if (gsl::at(char_speeds, i + 1).size() != num_grid_points) {
+      gsl::at(char_speeds, i + 1) = DataVector(num_grid_points);
+    }
+    gsl::at(rel_euler_char_speeds, i)
+        .set_data_ref(&gsl::at(char_speeds, i + 1));
```

Comment:
To expand: it would be helpful to document why the loops don't match the map between GRMHD <-> RelEuler given in the comment (aside: giving this map is 👍). If the loop upper bound was 4 (vs 5 as current) and the last special case was done outside the loop, then the code would look more like the comment and it would be easier to understand. I suggest either changing the code or adding comments to explain why it is being done the "hard" way.

Reply:
🎉 👍 for changing the code into something much easier to understand even though I only gave you a practically useless review comment.